Scanning workspace construction

ianbush opened the pull request:
A new class, `ScanningWorkspaceBuilder` has been added. This is meant to allow for easier construction of scanning workspaces.

`CreateSampleWorkspace` now supports building a sample scanning workspace using the builder.

A real loader using the `ScanningWorkspaceBuilder` is in progress, currently on the branch `d2b_loading`.

**To test:**

Code review. Check documentation is clear and complete.

Try creating a  scanning workspace with `CreateSampleWorkspace`. Check the dimensions, positions and angles are as expected.

Fixes #19198.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

ianbush committed:
Refs #19198 Initial attempt for scanning workspace construction

ianbush committed:
Refs #19198 First test passing for constructing workspace

ianbush committed:
Refs #19198 Renamed to ScanningWorkspaceBuilder, added tests

ianbush committed:
Refs #19198 Add positions and rotations to builder

ianbush committed:
Refs #19198 Added further checks and tests

ianbush committed:
Refs #19198 Add instrument angle method

ianbush committed:
Refs #19198 Set IndexInfo according to preference

ianbush committed:
Refs #19198 Remove class IndexInfo field

ianbush committed:
Refs #19198 Fix setting of detector and time indicies

ianbush committed:
Refs #19380 Add time index to detector table

CMake files updated to use the Types module

ianbush committed:
Refs #19380 use same formatting for time indexes as for detector IDs

ianbush committed:
Refs #19198 Merge branch 'master' into 19198_scanning_workspace_construction

Conflicts:
	Framework/Indexing/inc/MantidIndexing/IndexInfo.h

ianbush committed:
Refs #19198 Fix for tests after merge with IndexInfo changes

ianbush committed:
Refs #19198	Make use of default mapping in MatrixWorkspace

ianbush committed:
Refs #19380 Updated release notes

ianbush committed:
Refs #19198 Moved ScanningWorkspaceBuilder to DataObjects

ianbush committed:
Refs #19198 CreateSampleWorkspace has a scan workspace option

ianbush committed:
Refs #19198 Added simple test for creating a scanning workspace

ianbush committed:
Refs# 19198 Merge branch 'master' into 19198_scanning_workspace_construction

ianbush committed:
Refs #19198 Fix test failures

ianbush committed:
Refs #19198 Test for CreateSampleWorkspace

ianbush committed:
Refs #19198 Added documentation and tests

ianbush committed:
Refs #19198 Checks for setting positions and rotations

ianbush committed:
Refs #19198 Extract method for merging detector info

ianbush committed:
Refs #19198 Updated release notes for scanning

ianbush committed:
Merge branch 'master' into 19198_scanning_workspace_construction

OwenArnold commented on Framework/Algorithms/src/CreateSampleWorkspace.cpp:
```diff
@@ -159,6 +160,11 @@ void CreateSampleWorkspace::init() {
                   "The distance along the beam direction from the source to "
                   "the sample in M (default:10.0)");
+  declareProperty("NumScanPoints", 1,
+                  boost::make_shared<BoundedValidator<int>>(0, 360, true),
```
Where does the 360 bounds come from?

ianbush replied:
The detectors are rotated around the sample in 1 degree increments, hence the limit. Would ti be helpful to clarify this? It is only mentioned explicitly in the documentation.

OwenArnold replied:
It probably would be better to be clear on this "1 degree increments" aspect of the functionality

ianbush committed:
Refs #19380 Merge branch 'master' into 19380_time_indexes_for_detector_table

Conflicts:
	docs/source/release/v3.10.0/ui.rst

ianbush committed:
Refs #19380 Clang format

OwenArnold commented on Framework/DataObjects/src/ScanningWorkspaceBuilder.cpp:
```diff
@@ -0,0 +1,323 @@
+/**
+ * Supply a vector of vectors which contain positions. The inner vectors should
+ *contain the position for each time index, the outer vector the vector for each
+ *detector.
```
So should we not also compare things like the outer vector position size with outer rotation vector size? There may be some other examples for this. I didn't see that as part of the validation.

ianbush replied:
The outer vector size must always be the number of detectors, `m_nDetectors`. So they are ensured to be the same.

OwenArnold commented on Framework/DataObjects/inc/MantidDataObjects/ScanningWorkspaceBuilder.h:
```diff
@@ -0,0 +1,112 @@
+
+  void setIndexingType(const IndexingType indexingType);
+
+  API::MatrixWorkspace_sptr buildWorkspace();
```
Shouldn't this be a `const` method. I'm not clear why it wouldn't be.

OwenArnold commented on Framework/DataObjects/src/ScanningWorkspaceBuilder.cpp:
```diff
@@ -0,0 +1,323 @@
+  }
+}
+
+void ScanningWorkspaceBuilder::createTimeOrientedIndexInfo(
```
Should be `const` method

OwenArnold replied:
Also, since all these such functions are private, consider making these free functions and keeping the function declarations out of the header file. These can just live in an anonymous namespace.

ianbush replied:
All these methods make use of some member variables. I think this is possible only by passing in the extra information, so I am not sure if this will be so nice for code readability?

OwenArnold replied:
Ok. I think there might be some that take only use a single internal parameter. But i agree these `build...` ones might not make sense as non-members.

OwenArnold commented on Framework/DataObjects/src/ScanningWorkspaceBuilder.cpp:
```diff
@@ -0,0 +1,323 @@
+  ws.setIndexInfo(indexInfo);
+}
+
+void ScanningWorkspaceBuilder::createDetectorOrientedIndexInfo(
```
Should be `const` method

OwenArnold commented on Framework/Algorithms/src/CreateSampleWorkspace.cpp:
```diff
@@ -309,6 +320,36 @@ MatrixWorkspace_sptr CreateSampleWorkspace::createHistogramWorkspace(
+    timeRanges.push_back(double(i + 1));
+  }
+
+  builder.setTimeRanges(Kernel::DateAndTime(0), timeRanges);
```
I think it would be good to make the setters take r-value references. That way we can just move these temporaries rather than copying them.

OwenArnold commented on Framework/DataObjects/test/ScanningWorkspaceBuilderTest.h:
```diff
@@ -0,0 +1,509 @@
+    }
+  }
+};
+
```
No Performance test. I think tracking performance would probably be a good idea.

OwenArnold commented on Framework/DataObjects/inc/MantidDataObjects/ScanningWorkspaceBuilder.h:
```diff
@@ -0,0 +1,112 @@
+*/
+class MANTID_DATAOBJECTS_DLL ScanningWorkspaceBuilder {
+public:
+  enum class IndexingType { DEFAULT, TIME_ORIENTED, DETECTOR_ORIENTED };
```
What does `DEFAULT` mean?

ianbush replied:
Here `DEFAULT` means that `buildDefaultSpectrumDefinitions();` is called in `MatrixWorkspace::setIndexInfo()`, and no spectrum definitions are supplied.

I discussed this a little with @SimonHeybrock a while back, and it was decided not to promise any particular method for building the default spectrum definitions.

I wonder if my choice of names could be a little better here though...

SimonHeybrock commented on Framework/DataObjects/src/ScanningWorkspaceBuilder.cpp:
```diff
@@ -0,0 +1,323 @@
+void ScanningWorkspaceBuilder::createTimeOrientedIndexInfo(
+    MatrixWorkspace &ws) {
+  auto indexInfo = ws.indexInfo();
+  auto spectrumDefinitions = Kernel::make_cow<std::vector<SpectrumDefinition>>(
```
I would recommend simply using a vector here to reduce the hassle (move it when passing it to `IndexInfo`). The variant of `IndexInfo::setSpectrumDefinitions` that takes a `cow_ptr` is mainly for internal use by `MatrixWorkspace` and not required here.

SimonHeybrock commented on Framework/DataObjects/src/ScanningWorkspaceBuilder.cpp:
```diff
@@ -0,0 +1,323 @@
+void ScanningWorkspaceBuilder::createDetectorOrientedIndexInfo(
+    MatrixWorkspace &ws) {
+  auto indexInfo = ws.indexInfo();
+  auto spectrumDefinitions = Kernel::make_cow<std::vector<SpectrumDefinition>>(
```
See above.

SimonHeybrock commented on Framework/Indexing/inc/MantidIndexing/IndexInfo.h:
```diff
@@ -10,7 +10,11 @@
 namespace Mantid {
+
+using specnum_t = int32_t;
+using detid_t = int32_t;
```
The indexing module does not use those typedefs, there are type-safe types provided for this purpose, `SpectrumNumber` and `DetectorID`. Not sure why this was added to the header, maybe an oversight?

ianbush replied:
I am not sure how they got in there, this was indeed an oversight.

SimonHeybrock commented on docs/source/release/v3.10.0/framework.rst:
```diff
@@ -11,6 +11,7 @@ API
 
 - The default multiple file limit is now made facility dependent. It is 1000 for ILL, and 100 for all the others.
 - Frequency unit (GHz) included as an option to represent energy transfer.
+- Framework changes now mean scanning workspaces (workspaces with moving detectors) can be created. Currently this can be tested using a new option in :ref:`FilterEvents <algm-CreateSampleWorkspace-v1>`, by setting `NumScanPoints`.
```
Maybe we should also mention the limitations, mainly that the `InstrumentView` should not work, and algorithms may either fail or do nonsense?

ianbush replied:
Yes, we should, and that saving and loading will not give sensible results.

SimonHeybrock reviewed (commented):
Some comments below (have not yet looked through the tests):

OwenArnold commented on Framework/DataObjects/inc/MantidDataObjects/ScanningWorkspaceBuilder.h:
```diff
@@ -0,0 +1,112 @@
+
+  HistogramData::Histogram m_histogram;
+
+  boost::shared_ptr<const Geometry::Instrument> m_instrument;
```
Just an idea - for member variables that are mandatory, one way you might wish to avoid logical misuse errors is to wrap them as `boost::optional`. That way there is no danger of blindly accepting default initialised values if setters have not been called. You can always check that these have been initialised prior to use via `boost::optional::operator*()`

ianbush commented on Framework/DataObjects/src/ScanningWorkspaceBuilder.cpp:
```diff
@@ -0,0 +1,323 @@
+ * @param instrumentAngles a vector of angles, the size matching the number of
+ *time indexes
+ */
+void ScanningWorkspaceBuilder::setInstrumentAngles(
```
I am in two minds about this too. On balance I think this method is general enough to be here, as all real cases for scanning instruments at ILL and MLZ rotate around the sample position in such a way.

SimonHeybrock replied:
Would the code be any more complex if this was not a member?

ianbush commented on Framework/DataObjects/src/ScanningWorkspaceBuilder.cpp:
```diff
@@ -0,0 +1,323 @@
+  switch (m_indexingType) {
+  case IndexingType::DEFAULT:
+    outputWorkspace->setIndexInfo(
+        Indexing::IndexInfo(m_nDetectors * m_nTimeIndexes));
```
Yes, you are right, this is a bit confusing at the moment. For now I think it makes sense to ensure the number of detectors in the instrument and `nDetectors` are the same.

I think moving the instrument into the constructor might be better, then `nDetectors` does not need to be given separately.

ianbush committed:
Refs #19198 Minor fixes from PR review

ianbush committed:
Refs #19198 Pass by value and move in builder

ianbush committed:
Refs #19198 Move instrument setting to constructor

ianbush committed:
Refs #19198 Some further fixes from code review

ianbush committed:
Refs #19198 Added performance test

ianbush committed:
Refs #19198 Removeded redundante instrument check

ianbush committed:
Refs #19198 Renamed instrument angles to relative rotations

SimonHeybrock commented on Framework/DataObjects/inc/MantidDataObjects/ScanningWorkspaceBuilder.h:
```diff
@@ -54,50 +54,50 @@ namespace DataObjects {
-  void setRotations(const std::vector<std::vector<Kernel::Quat>> &rotations);
-  void setInstrumentAngles(const std::vector<double> &instrumentAngles);
+  void setPositions(const std::vector<std::vector<Kernel::V3D>> positions);
+  void setRotations(const std::vector<std::vector<Kernel::Quat>> rotations);
```
If you want to move from these they should not be `const` (which is ok, since you are no passing by value).

SimonHeybrock commented on Framework/DataObjects/src/ScanningWorkspaceBuilder.cpp:
```diff
@@ -17,24 +17,14 @@ using namespace Mantid::Indexing;
+    : m_nDetectors(instrument->getNumberDetectors()),
+      m_nTimeIndexes(nTimeIndexes), m_nBins(nBins), m_instrument(instrument),
+      m_histogram(BinEdges(nBins + 1, LinearGenerator(1.0, 1.0)),
                   Counts(std::vector<double>(nBins, 0.0))),
```
You can simply pass `nBins, 0.0`, `Counts` has the same constructors as `std::vector`.

SimonHeybrock commented on Framework/DataObjects/src/ScanningWorkspaceBuilder.cpp:
```diff
@@ -44,13 +34,13 @@ void ScanningWorkspaceBuilder::setInstrument(
-    const HistogramData::Histogram &histogram) {
+    const HistogramData::Histogram histogram) {
   if ((histogram.counts().size() != m_nBins) ||
       (histogram.binEdges().size() != m_nBins + 1))
```
Use `Histogram::size()`. `Histogram` guarantees that the bin edges are always one more than the number of bins, no need to checking.

ianbush committed:
Refs #19198 Relative rotations can take position and axis

ianbush committed:
Refs #19198 Minor fixes

ianbush committed:
Merge branch 'master' into 19380_time_indexes_for_detector_table

ianbush committed:
Refs #19380 Fix release notes

SimonHeybrock commented on Framework/DataObjects/src/ScanningWorkspaceBuilder.cpp:
```diff
@@ -243,30 +235,31 @@ void ScanningWorkspaceBuilder::buildPositions(
+void ScanningWorkspaceBuilder::buildRelativeRotationsForScans(
     DetectorInfo &outputDetectorInfo) const {
   for (size_t i = 0; i < outputDetectorInfo.size(); ++i) {
     for (size_t j = 0; j < outputDetectorInfo.scanCount(i); ++j) {
```
This is what I referred to on Slack (a bit cryptically maybe): If you would apply these rotations before merging the `DetectorInfos`, you could simply rotate a parent of the detectors via `DetectorInfo::setRoation(IComponent *, ...)`. That should also be much faster, since it uses `Eigen` directly internally. In principle rotating the full instrument would work, but that would also rotate monitors etc., so you'd have to figure out how to handle those.

SimonHeybrock commented on Framework/DataObjects/src/ScanningWorkspaceBuilder.cpp:
```diff
@@ -0,0 +1,312 @@
+  for (size_t i = 0; i < outputDetectorInfo.size(); ++i) {
+    for (size_t j = 0; j < outputDetectorInfo.scanCount(i); ++j) {
+      auto position = outputDetectorInfo.position({i, j});
+      const auto rotation = Kernel::Quat(m_instrumentAngles[j], m_rotationAxis);
```
Can you double check what this does? `Eigen::Quaternion` requires the axis to be normalized, otherwise you get garbage (a quaternion that is not representing a rotation), not sure about `Kernel::Quat`.

ianbush replied:
The rotation axis is normalised for you, see - https://github.com/mantidproject/mantid/blob/master/Framework/Kernel/src/Quat.cpp#L117.

SimonHeybrock commented on Framework/DataObjects/src/ScanningWorkspaceBuilder.cpp:
```diff
@@ -0,0 +1,312 @@
+      rotation.rotate(position);
+      position += m_rotationPosition;
+      outputDetectorInfo.setPosition({i, j}, position);
+      outputDetectorInfo.setRotation({i, j}, rotation);
```
You are dropping the previous rotation here, i.e, rotation is treated as an absolute value rather than an offset.

ianbush committed:
Refs #19198 Make sure old detector rotations are not ignore

SimonHeybrock commented on Framework/DataObjects/test/ScanningWorkspaceBuilderTest.h:
```diff
@@ -320,6 +328,58 @@ class ScanningWorkspaceBuilderTest : public CxxTest::TestSuite {
+            int(nDetectors), int(nBins));
+    auto &instDetInfo = instWS->mutableDetectorInfo();
+
+    Quat rotation = Quat(10.0, V3D(0, 1, 0));
```
If both the original rotation and the updated one are around the same axis you are not catching a wrong rotation order (remember, rotation matrices to not commute!).

ianbush committed:
Refs #19198 Orthogonal rotation to check order rotations are applied

ianbush committed:
Refs #19198 Work around atan2 problem correctly

SimonHeybrock commented:
I am happy with the code now, tests have been added, and I think despite scanning being a new feature there is not really any risk of breaking anything with this since the changes do not affect existing code/workflows (in case anyone is concerned about doing this just before a release).

:shipit:

ianbush committed:
Refs #19380 Use Types as header only

martyngigg commented:
@ianbush There seems to be a failing test on Windows. Do you have access to a Windows machine?

ianbush commented:
@martyngigg Yes, I do, I will take a look at the failing test this morning. It is a slightly worrying difference in results...

ianbush committed:
Refs #19198 Change order of angles to avoid rounding issue

ianbush committed:
Refs #19198 Fixed test for Y value

ianbush committed:
Merge branch '19380_time_indexes_for_detector_table' into 19198_scanning_workspace_construction

martyngigg commented:
@ianbush The final failure is a real failure in the doc test [example code](/mantidproject/mantid/blame/master/docs/source/algorithms/RecordPythonScript-v1.rst#L69). The new parameter on the algorithm has caused the output comparison to fail.

ianbush committed:
Refs #19198 Fix for RecordPythonScript doc test

ianbush committed:
Refs #19198 Revert "Merge branch '19380_time_indexes_for_detector_table'

into 19198_scanning_workspace_construction"

This reverts commit 29a31587c9900839f97c753be752a13bb0344c53, reversing
changes made to d474e22a0826b23f4a8025f5d9c57a39c15ad9ee.

ianbush commented:
Please see new PR #19606.
